docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,286 @@
+  same gateway it is allowed to process the message. If not, we should inform the previously owning
+  LNS to drop the connection and allow the new LNS to process the message. This was not preferred as
+  it has more complexity for unclear results.  
+
+### Main data message flow
+
+Currently, if LoRaDevice is not in the LNS cache, we search on the Function for all devices that
+have that DevAddr. Then we get all their twins which could result in a connection switch.
+
+#### Decision
+
+After loading all the twins for the LoRaDevice(s), we close the connections immediately.
+This is equivalent to how we [handle Join requests](#handling-of-join-requests)
+
+#### Alternative
+
+- Currently the DeviceGetter.GetDevice returns a list of devices that match the provided DevAddr. We
+  considered changing the Function to return a single device instead of a list and only load the
+  twin for this one on the LNS side. For this the Function should perform the Mic computation which
+  means that we would need to send the payload to the cloud. This is a deal breaker for us.
+
+#### Further processing
+
+This section uses this topology:
+
+```mermaid
+flowchart LR;
+    Device-->LBS1-->LNS1-->Function;
+    Device-->LBS2-->LNS2-->Function;
+    LNS1-->IoTHub;
+    LNS2-->IoTHub;
+```
+
+where Device sends data message A and then B.
+
+Here is a rundown of what should happen assuming both LNSs have a fresh twin (via background refresh
+or via fetching it using DeviceGetter.GetDevice). Changes are marked in **bold**:
+
+1. Device sends first data message A.
+1. We assume that LNS1 gets the message first. **LNS1 checks against
+   its in-memory dictionary DevEui -> flag** and since it has not seen this DevEui before contacts the Function.
+1. The Function hasn't seen this DevEui either and therefore does not have an assigned LNS for it
+  yet. LNS1 wins the race and gets immediately a response and processes the message upstream.
+1. LNS2 eventually receives message A, **checks its local dictionary** and also contacts the
+   FunctionBundler immediately since it does not have prior info about this DevEui.
+1. The Function responds to LNS2 that it lost the race to process this message.
+1. Since deduplication strategy is Drop, LNS2 drops the message immediately, therefore no
+   connection to Iot Hub is opened and only LNS1 has the connection to Iot Hub. **LNS2 notes in
+   memory that it was the losing gateway for this DevEui**.
+1. When message B gets send (with a higher frame counter*), assuming that this time LNS2 gets it
+   first it **checks again its local dictionary it's not the preferred LNS for this device and
+   therefore delays itself X ms before contacting the FunctionBundler**.
+   - Here we do *not* want to simply drop the message as LNS1 might not be available anymore (due to
+     a crash, device not in range etc).
+   - Since it was the losing LNS previously it should **make sure that it operates on a fresh twin**
```

Comment:
this is relevant only for the fcnt drift we discussed. We will keep the background refresh of the twin also on the not owning device to ensure it's refreshed periodically. The connection will be opened only temporarily and closed after the refresh.

Reply:
correct, removed this now (104740f )